### PR TITLE
[ENH] Add auto information

### PR DESCRIPTION
**Pull request by vferat**
Adapt code from https://github.com/Frederic-vW/AIF-PAIF as discussed in [this issue](https://github.com/Frederic-vW/AIF-PAIF/issues/3) and by emails.

### CODE REVIEW

**Comment by vferat** [comment]
Here, I reuse the parameter `ignore_self` introduced in the #91. I find that the name  isn't very explicit, especially outside the context of transition probabilities ( "transition to itself"). I think it would be good to have a generic and explicit name that would indicate when an analysis considers segmentation with state repetition (AABBBC) or without (ABC).

@mscheltienne  do you think it's a big problem to change the variable name (for backwards compatibility) ? Maybe we could introduce it as an alias for the transition module.
Also, do you have any suggestion for a more explicit name ?  `ignore_repeated_states`, `ignore_repeation`?

File: pycrostates/segmentation/entropy.py
Comment (mscheltienne) [inline pycrostates/segmentation/entropy.py:206]: I fixed the type-hints/docstring for those 3 arguments. Previously, it was:

```
state_to_ignore: Optional[Union[int, None]]
ignore_self: Optional[bool]
log_base: Optional[Union[float, str]]
```

The first one, I think it's a tpyo.
The 2 others, if None is provided one of the checker would raise.

File: pycrostates/segmentation/entropy.py
Comment (mscheltienne) [inline pycrostates/segmentation/entropy.py:283]: Function called only once in `excess_entropy_rate`. Any particular reason to split those 5 lines in a separate function? Else I would prefer to keep the code short and to move those 5 lines below in the `excess_entropy_rate` function.

**Review by mscheltienne — COMMENTED** [review]
I went through it once today, looks good. A couple of minor comments.
For the `ignore_self` variable, let's discuss an alternate. I think I would go towards the naming of `itertools`, IIRC, it's `with_replacement`. 
We can deprecate the existing argument, release; and then remove it when we release the next version when MNE 1.6 rolls out, in about 6 months.

File: docs/source/dev/changes/latest.rst
Comment (mscheltienne) [inline docs/source/dev/changes/latest.rst:28]:
```suggestion
- Add :func:`~pycrostates.segmentation.entropy`, :func:`~pycrostates.segmentation.excess_entropy_rate`, :func:`~pycrostates.segmentation.auto_information_function` and :func:`~pycrostates.segmentation.partial_auto_information_function` (:pr:`112` by `Victor Férat`_ and `Frederic von Wegner`_)
```

**Comment by vferat** [comment]
> Function called only once in `excess_entropy_rate`. Any particular reason to split those 5 lines in a separate function? Else I would prefer to keep the code short and to move those 5 lines below in the `excess_entropy_rate` function.

I reworked the API so:
- All **private** function are defined for input `labels: NDArray[int]` and all **public** functions are defined for input` segmentation: Segmentation`
-  `state_to_ignore` is removed from all **public** functions (hard coded `state_to_ignore=-1`).
- the ignore_self logic is defined in the  **public** functions

Overall, **private** functions are _generic_ functions made to work with `np.array` and custom `state_to_ignore` while **public** ones are made to work only within pycrostates workflow ( `Segmentation `input and `state_to_ignore=-1`)

**Comment by mscheltienne** [comment]
I'll have a look at the code, but the description sounds good!

**Comment by vferat** [comment]
> I went through it once today, looks good. A couple of minor comments. For the `ignore_self` variable, let's discuss an alternate. I think I would go towards the naming of `itertools`, IIRC, it's `with_replacement`. We can deprecate the existing argument, release; and then remove it when we release the next version when MNE 1.6 rolls out, in about 6 months.

 `with_replacement` is not explicit enough in my opinion and the itertools' use case is a bit different.
 What about `remove_repetitions`, `ignore_length` or `ignore_duration `?

**Comment by vferat** [comment]
```
C:\Users\ferat\Documents\Github\pycrostates\docs\source\api\segmentation.rst:21:<autosummary>:1: ERROR: Too many autonumbered footnote references: only 0 corresponding footnotes available.
C:\Users\ferat\Documents\Github\pycrostates\docs\source\api\segmentation.rst:21:<autosummary>:1: ERROR: Unknown target name: "footcite-von2018partial".
```

@mscheltienne Any idea ?

**Comment by mscheltienne** [comment]
Slight commit to change the render, in[1] is a bit weird. I think it's better with an author name as in the microstates glossary/description:

> .. sidebar:: Looking for more information?
> 
>     If you are looking to learn more about microstates we recommend you to read
>     the following review: :footcite:t:`MICHEL2018577`.

**Comment by vferat** [comment]
> I tried to add a footbibliography (see above), the footbibliography is rendered correctly but the link in the autosummary table is still dead. The problem is that the first line of the docstring is used for the autosummary description, and somehow that does not play well with the bibliography links.

Ican also remove the citation from the autosummary  and add it to the complete function description instead

**Comment by mscheltienne** [comment]
> I can also remove the citation from the autosummary

Yes, that works too. I'm just a bit annoyed it doesn't work 😅

**Comment by vferat** [comment]
> Also, while I think about it, I think we can drop the `_function` at the end of `auto_information_function` and `partial_auto_information_function`.

A quick note here: the term _function_ is used in the sense of a mathematical function _f(x)_  as the output of the python function is _x,(f(x)_
Now that it's said, we can still drop the word function, let me know what you think

**Comment by mscheltienne** [comment]
Oh, OK, fine with it then.
It would help to have the type-hints for the `return`.

**Comment by mscheltienne** [comment]
> I can also remove the citation from the autosummary

I think that's going to be the solution, I'm not finding an example with a citation in the summary line of a docstring.

**Comment by vferat** [comment]
I deprecated `ignore_self ` in favour of  `ignore_repetitions`.

- **New** functions only use  `ignore_repetitions`.

- **Public** functions previously using the  `ignore_self `  parameter are modified as follow:
  -  change default value of `ignore_self` from `True `to  `None` .
  - add parameter  `ignore_repetitions` with  default value  `True` .
  - a warning is emtitted if `ignore_self` is not `None`
  - if `ignore_self` is not `None`,  `ignore_repetitions` is set to `ignore_self`
  - `ignore_self` docstring is now a deprecation warning

    Should we handle the case where users set both  `ignore_repetitions` and  `ignore_self` ?


- **Private**  functions previously using the  `ignore_self ` now use the `ignore_repetitions` instead.

**Comment by mscheltienne** [comment]
I'll try to look this week, that should be the last blocker before the next release.

**Review by mscheltienne — APPROVED** [review]
That looks good to me. I pushed a couple of fixes and a better deprecation using `warnings.warn(DeprecationWarning, ...)`. The CI failures will be fixed by #124

@vferat Happy to merge and to improve the tutorial in a subsequent PR?

**Comment by vferat** [comment]
@Frederic-vW who is behind the code and research implemented in this **PR** is reviewing the changes and is also writing a more detailed tutorial.
 I'd prefer to wait for his feedback before merging these new features into main.
 
 If we have to do a release soon, I think we can do it without this PR and integrate them later.

**Comment by mscheltienne** [comment]
OK, we need to cut a release right before MNE 1.6, somewhere in November.

**Comment by mscheltienne** [comment]
MNE1.6 will be out this week or next week, I'll cut a release now.

**Comment by vferat** [comment]
All good for me, we'll merge this PR for the next one

**Comment by mscheltienne** [comment]
Oh and I'll move the argument deprecation from this PR to `main`, to get the deprecation released.

**Comment by mscheltienne** [comment]
I took care of the conflicts and removed the deprecation from this PR.

**Comment by vferat** [comment]
@mscheltienne ready for review !

**Pull request merged**
